service/controller/v3/resource/chart/desired.go

```diff
@@ -65,3 +73,47 @@ func (r *Resource) getConfigMapValues(ctx context.Context, customObject v1alpha1
 
 	return chartValues, nil
 }
+
+func (r *Resource) getSecretValues(ctx context.Context, customObject v1alpha1.ChartConfig) (map[string]interface{}, error) {
+	secretValues := make(map[string]interface{})
+
+	if key.SecretName(customObject) != "" {
+		secretName := key.SecretName(customObject)
+		secretNamespace := key.SecretNamespace(customObject)
+
+		secret, err := r.k8sClient.CoreV1().Secrets(secretNamespace).Get(secretName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return nil, microerror.Maskf(notFoundError, "secret '%s' in namespace '%s' not found", secretName, secretNamespace)
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		// TODO: fix this "secret.json" name somewhere and access it in release-operator.
+		secretData := secret.Data["secret.json"]
+		if secretData != nil {
+			err = json.Unmarshal(secretData, &secretValues)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+		}
+	}
+
+	return secretValues, nil
+}
+
+func union(a, b map[string]interface{}) (map[string]interface{}, error) {
+	if a == nil {
+		return b, nil
+	}
+
+	for k, v := range b {
+		_, ok := a[k]
+		if ok {
```

Comment:
`a[k]` tries to get the value with key `k` from map `a`. `ok` is true if we were able to get the value from `a`. So the retrieval was "ok".

I this context we don't want to have the same key in both `a` and `b`. Since we get the key `k` from the `range b` it is present in `b` and if `ok` is true then it is also present in `a`.

Does that clear it up a bit?

Reply:
1. In go you return errors, not throw them (I have allergy to that sorry :D)
2. This `ok` name is just a convention - you can think of it as "contains". Don't confuse it with actual logic.